nemo/collections/asr/parts/preprocessing/perturb.py

```diff
@@ -1055,25 +1243,58 @@ def __init__(self, manifest_path: str, tar_filepaths: Union[str, List[str]], shu
         else:
             logging.info("WebDataset will not shuffle files within the tar files.")
 
-        self.audio_dataset = self.audio_dataset.rename(audio='wav', key='__key__').to_tuple('audio', 'key')
-        self.audio_iter = iter(self.audio_dataset)
+        self.audio_dataset = (
+            self.audio_dataset.rename(audio='wav;ogg;flac', key='__key__')
+            .to_tuple('audio', 'key')
+            .pipe(self._loop_offsets)
+        )
 
     def __len__(self):
         return len(self._manifest)
 
+    def _loop_offsets(self, iterator):
+        """This function is used to iterate through utterances with different offsets for each file.
+        """
+
+        class TarredAudioLoopOffsets:
+            def __init__(self, collection):
+                self.iterator = iterator
+                self.collection = collection
+                self.current_fn = None
+                self.current_bytes = None
+                self.offset_id = 0
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
```

Comment:
I thought we'd move to iterable pattern ? (Ie no `__next__`?

Reply:
This function I took directly from _TarredAudioToTextDataset located here: https://github.com/NVIDIA/NeMo/blob/41fcf4daccee9aa5082431f4ae89c76fc685eac9/nemo/collections/asr/data/audio_to_text.py#L903

It seems for the map/filter functions used there, it uses the Iterator pattern, but those are filter function helpers fed to the map inside Webdataset, so they probably need to be Iterators.

Reply:
Ok cool